### DISCUSSION TITLE
Creating new PFT

### DISCUSSION THREAD

**Discussion by evalieungh** [NGEET/fates#925] [General]
Hi all, 
This is meant as a newbies discussion for working with/creating PFTs.

We have a master student (connected to the EMERALD project in Norway) that wants to create a new PFT for Salix shrubs in FATES and see how the model behaves differently with and without it on a local or regional scale. We're still in the process of designing the project but I thought I'd ask right away how we might go about doing this in case we need to change the project design accordingly. 

### How and where are PTFs defined?
Is it correct that the PFTs are defined in /parameter_files/fates_params_default.cdl ? In other issues I've read talk about PFT parameter files (as in one file per PFT?) but haven't found anything like that in the  NGEET/fates repository.

In the fates_params_default.cdl file:
Line 10 seems to specify the number of PFTs used (default = 12), and their names are listed in lines 606-618. Then follows a long list of parameters with 12 values - I assume one per PFT. 
Some parameters (?), like fates_hydr_avuln_node (lines 735-739), have more numbers than PFTs. Why? 

To make a new PFT and run the model, is it enough to add a line with the name, and add a number in the appropriate order in the parameters in the fates_params_default.cdl file? Or do other files need to be changed too? 
Alternatively, is it easier to co-opt an existing PFT, e.g. a tropical one that should not be in our study area, and change the numbers? 

### Parameter descriptions 
The parameters listed in fates_params_default.cdl are not immediately understandable. Is there a description of the parameter somewhere, e.g. a link between the parameter names and some documentation (or literature references) of their meaning and where the values come from? 

Any guidance is greatly appreciated!

**Comment by ckoven** [comment]
Hi Eva,
What is needed to define a new PFT is only to add that PFT in the parameter file.  Probably the easiest way to do that is to use the tool https://github.com/NGEET/fates/blob/master/tools/FatesPFTIndexSwapper.py to make an copy of one of the existing PFTs via a command such as `FatesPFTIndexSwapper.py --pft-indices=1,1,2,3,4,5`, and then change the relevant parameters that make the new PFT different from existing ones.  Note that the PFT names aren't actually used in the code, they are just there to help keep track of the intention behind a given PFT definition.

The extra lines in variables such as fates_hydr_avuln_node are due to those variables having a more complex dimension.  In the case of fates_hydr_avuln_node, it is dimensioned (fates_hydr_organs, fates_pft).  You can look this up in the part of the cdl file where the variables are defined, here: https://github.com/NGEET/fates/blob/master/parameter_files/fates_params_default.cdl#L150

We are working on making the parameter definitions more transparent.  A goal is to make a gigantic table in the tech note that points each parameter name to the relevant variable in the tech note that might have further info, but we haven't gotten there quite yet.  Any help in moving in that direction would be great.

**Comment by jenniferholm** [comment]
Hi Eva, 

Charlie's answer was right on for creating a new PFT in the PFT parameter file, and providing the link to the current parameter descriptions. Like Charlie mentioned you would make a copy of an existing PFT (probably one of the existing deciduous shrubs), and add on as the 13th PFT. 

How does your Salix shrub differ from the two deciduous shrubs currently in FATES? The "colddecid" shrub is a combination of shrubs that exist in both temperate and boreal cold regions, and is seasonally deciduous. The "hydrodecid" shrub is deciduous due to hydraulic stress. 

I also wanted to chime in because we have a postdoc arriving soon who will work on adding arctic shrubs into FATES. I'd love to stay in touch, and maybe we can learn from each other, and help when needed. 

Thanks, Jennifer

**Comment by evalieungh** [comment]
Thanks @ckoven and @jenniferholm for clarifying! Using the IndexSwapper tool seems doable. 

I'm not sure to what extent the cold deciduous shrubs PFT already present differs from our salix systems, but I think it will be interesting to find out if a 'narrower' PFT definition for Norwegian mountains will predict occurrence of the PFT more accurately (against national field-based estimates). If you are working on a similar line we should definitely stay in touch. My email is e.l.eriksen[at]nhm.uio.no. 

We are hoping to measure some of the parameters in the field, at localities along/above the treeline, and might be able to contribute with some parameterisation improvements to the standard PFTs at the end of the project. 

I'll make sure we document our method thoroughly, so it will be easier to use for other beginners! Charlie, defining the parameters is something we will have to do in order to measure them so I hope we can add a bit to the documentation there as well.

Should we close this issue, or keep it open for the inevitable new questions when we actually start trying to do this?

**Comment by rosiealice** [comment]
Hi @evaleriksen... 

A couple more things to add on top of @ckoven and @jenniferholm's.

1. I'd not treat the 'default' PFT file as being very fixed at the moment. We are working towards a better global representation, but if i were you, I'd test the new PFT first in isolation, and then perhaps with a small set of relevant competing PFTs, to avoid odd competitive outcomes that we see a lot in testing with all the plant types at once.  

2. Charlies script that he described will automatically make a file with the number of PFTs that you want/need, and then the model will run with all of those PFTs, so you don't need to change anything manually.  

3. There is another script that @ckoven  wrote
https://github.com/NGEET/fates/blob/master/tools/modify_fates_paramfile.py
 that changes individual values in the PFT file, such that you can start with, say, the default file, pick out the PFTs you want, and then change the relevant values, all in one go and with a record of what you did (which is coool). That also means that if the default file changes (we are, for example, often adding new parameters) you can quickly make another file that works. 

I put a simple version of a script that takes two cold-decid trees and modifies their 'fates_phen_mindayson' parameter here:
https://github.com/rosiealice/jupyter_notes/blob/master/python_scripts/pft_mod_for_CLD

To make this work, you'll need to have a '.nc' version of the default parameter file in the same directory as those two python scripts. To get that, you can either use 'ncgen' on the '.cdl' file like this
_ncgen -o fates_params_default.nc fates_params_default.cdl_ 
(the opposite is using _ncdump fates_params_default.nc > fates_params_default.cdl_ )

or you can find the version that FATES is using already, (it's generated at some point during configuring & building and the path to it is listed under 'fates_paramfile' in the 'lnd_in' file in the 'run' directory where your model output is generated. 

4. I assume your salix PFT will be cold deciduous, in which case, you'll probably want to check #574, wherein we fixed some strange issues that had crept in to the phenology.   There is currently a pull request up (https://github.com/NGEET/fates/pull/579) to fix these, so I'd either use what is there or wait until it gets pulled into the master code once it's all been tested etc. 

5. Will you want the phenological triggers to be different for the salix than the other PFTs? At the moment, phenology is all determined at the 'site' level, meaning that all the plants have the same cold or drought phenological status. One thing I've been thinking about especially for drought is to change this to having different timings for different PFTs, so I'm interested in whether that would be on your radar as well. 

Good to see you getting into this. I just saw your reply above, but I think this is all still relevant!

**Comment by evalieungh** [comment]
Hi @rosiealice , thanks for these additions and good advice. 

1. Is it now possible to run FATES with only one PFT? I thought there had to be at least two? But we will definitely do some testing to see if the new PFT is well-behaved or not... 

Thanks for the links to the scripts, I'm sure they will be useful. The master student has just started and is busy with courses for a few more months, so I hope #579 will be pulled before we run into that problem.

5. Phenology is not currently part of our thinking in this project, but it does seem like something that should differ between PFTs realistically. It would be fun to include, but I think it would be too complicated for a master thesis.

**Comment by rosiealice** [comment]
Cool. Yes, whatever it was that prevented us suing only one PFT has now gone away :) Let us know when you get started if you need any help.

**Comment by rgknox** [comment]
closing due to time elapsed, feel free to re-open and continue discussion though!

**Comment by wwieder** [comment]
I'm curious what progress has been made on this work?  It seems like a few teams were going to start exploring the Salix question, but what came of the efforts? Are there papers, branches, or parameterizations for arctic shrubs that have been developed and tested? What can we do to keep working developing this regional parameterization (specifically for alpine veg)?

**Reply by jenniferholm** [reply]
Hi will,  I'm not sure what has become of the Salix PFT FATES work, and it would be nice to hear an update if anything has progressed (or maybe I'm just out of the loop).  

We've worked a little bit on sensitivity analysis and parameterization for arctic shrubs (based on some field trait data from the Seward peninsula in Alaska, but only a little constrained data at this point).  I'm not sure how different alpine vs. arctic shrubs would be, do you? But at least it would be a start.  
Any, short answer to your question, yes we have some beginning work on arctic shrubs and I'll share some of that insight! _After_ I check in with Yanlan Liu, and get her ok, who is the main owner of this analysis, and is working on a publication with this data.

**Reply by alistairrogers** [reply]
Hi Wil, 
I’ve published data on photosynthetic parameters for Salix pulchra from Utqiagvik (formerly Barrow), AK, and have data on other Salix species and Alnus from the Seward Peninsula, AK. These data include Vcmax25, Jmax25, LMA and Na. The wider NGEE Arctic project has much more data on plant traits for these species that could be used for parameterization. Jennifer should be able to help you identify data sets and dataset authors through the NGEE Arctic project data repo. Work is ongoing so you can expect more shrub related parameterization to be available in the next 12-18 months.
Alistair

**Reply by evalieungh** [reply]
An update from my side as well. The master student I mentioned sadly decided to go in another direction (thanks to covid and my own probably too optimistic plans), so we never got beyond the planning stage. I can also mention that some colleagues in Bergen (++), Norway just got a new grant to study dwarf shrubs in the next few years. I think @huitang-earth is involved. Probably not focused on Salix, but on other shrubs. 
Great to hear others are working on Salix parameters, though!

**Comment by glemieux** [comment]
Moved to discussions since it seemed appropriate.

**Comment by evalieungh** [comment]
For any newbies wanting to do single-point simulations with PFT alterations, I can also mention our new ['land sites platform'](https://github.com/NorESMhub/noresm-land-sites-platform) that has a graphical user interface where PFTs can be removed and a subset of the FATES parameters can be modified. The GUI doesn't support adding new PFTs, but the parameter file can be modified behind the scenes for those who are a bit familiar with the model. Suitable for those without access to Cheyenne, for instance. 
Tagging @huitang-earth and @lasseke FYI

**Reply by evalieungh** [reply]
NB we have still not submitted our manuscript describing the LSP, but the software is functional and open.
